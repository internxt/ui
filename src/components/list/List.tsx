import _ from 'lodash';
import React, { ReactNode, useEffect, useRef, useState } from 'react';
import ListItem from './ListItem';
import ListHeader, { HeaderProps } from './ListHeader';
import useHotkeys from '../../hooks/useHotKeys';
import { SkeletonLoader } from '../skeletonLoader';
import { InfiniteScroll } from '../infiniteScroll';
import { MenuItemsType } from '../menu/Menu';

export interface ListProps<T, F> {
  header: HeaderProps<T, F>[];
  checkboxDataCy?: string;
  items: T[];
  itemComposition: Array<(props: T) => JSX.Element>;
  selectedItems: T[];
  onClick?: (props: T) => void;
  onDoubleClick?: (props: T) => void;
  onEnterPressed?: (props: T) => void;
  onSelectedItemsChanged: (changes: { props: T; value: boolean }[]) => void;
  isLoading?: boolean;
  forceLoading?: boolean;
  skinSkeleton?: Array<JSX.Element>;
  emptyState?: ReactNode;
  onNextPage?: () => void;
  onOrderByChanged?: (value: { field: F; direction: 'ASC' | 'DESC' }) => void;
  orderBy?: { field: F; direction: 'ASC' | 'DESC' };
  hasMoreItems?: boolean;
  menu?: MenuItemsType<T>;
  displayMenuDiv?: boolean;
  className?: string;
  keyboardShortcuts?: Array<'selectAll' | 'unselectAll' | 'multiselect' | Array<'delete' & (() => void)>>;
  disableKeyboardShortcuts?: boolean;
  disableItemCompositionStyles?: boolean;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  keyBoardShortcutActions?: {
    onShiftFKeysPressed?: () => void;
    onRKeyPressed?: () => void;
    onBackspaceKeyPressed?: () => void;
  };
}

/**
 * List component
 *
 * @property {HeaderProps<T, F>[]} header
 * - Array of headers for the list. Each header defines the properties of the columns displayed.
 *
 * @property {string} [checkboxDataCy]
 * - Optional `data-cy` attribute used for testing checkboxes in the list.
 *
 * @property {T[]} items
 * - Array of items to be displayed in the list. Each item corresponds to the data for each row.
 *
 * @property {(props: T) => JSX.Element}[] itemComposition
 * - Array of functions that return JSX elements for rendering each item in the list.
 *
 * @property {T[]} selectedItems
 * - Array of selected items.
 *
 * @property {(props: T) => void} [onClick]
 * - Optional callback triggered when an item is clicked.
 *
 * @property {(props: T) => void} [onDoubleClick]
 * - Optional callback triggered when an item is double-clicked.
 *
 * @property {(props: T) => void} [onEnterPressed]
 * - Optional callback triggered when the Enter key is pressed on an item.
 *
 * @property {(changes: { props: T; value: boolean }[]) => void} onSelectedItemsChanged
 * - Callback triggered when the selection state of the items changes.
 *
 * @property {boolean} [isLoading]
 * - Optional flag indicating if the list is loading. If true, a loading state will be shown.
 *
 * @property {boolean} [forceLoading]
 * - Optional flag to force the loading state, even if no items are loading.
 *
 * @property {JSX.Element[]} [skinSkeleton]
 * - Optional array of skeleton elements to be displayed while the list is loading.
 *
 * @property {ReactNode} [emptyState]
 * - Optional content to display when there are no items in the list and no loading state.
 *
 * @property {() => void} [onNextPage]
 * - Optional callback triggered when the user scrolls to load the next page of items.
 *
 * @property {({ field: F; direction: 'ASC' | 'DESC' }) => void} [onOrderByChanged]
 * - Optional callback triggered when the user changes the sorting order of the list.
 *
 * @property {{ field: F; direction: 'ASC' | 'DESC' }} [orderBy]
 * - Optional object specifying the current sorting state of the list.
 *
 * @property {boolean} [hasMoreItems]
 * - Optional flag indicating if there are more items to load.
 *
 * @property {MenuItemsType<T>} [menu]
 * - Optional menu items to be displayed for each item in the list.
 *
 * @property {boolean} [displayMenuDiv]
 * - Optional flag to control whether the menu is displayed as a separate div.
 *
 * @property {string} [className]
 * - Optional CSS class name to apply additional styles to the list container.
 *
 * @property {Array<'selectAll' | 'unselectAll' | 'multiselect' | Array<'delete' & (() => void)>>} [keyboardShortcuts]
 * - Optional array of keyboard shortcut actions to be handled.
 *
 * @property {boolean} [disableKeyboardShortcuts]
 * - Optional flag to disable keyboard shortcuts for the list.
 *
 * @property {boolean} [disableItemCompositionStyles]
 * - Optional flag to disable custom styles for item composition.
 *
 * @property {() => void} [onMouseEnter]
 * - Optional callback triggered when the mouse enters the list.
 *
 * @property {() => void} [onMouseLeave]
 * - Optional callback triggered when the mouse leaves the list.
 *
 * @property {
 *              { onShiftFKeysPressed?: () => void; onRKeyPressed?: () => void; onBackspaceKeyPressed?: () => void }
 *           } [keyBoardShortcutActions]
 * - Optional object with custom actions for specific keyboard shortcuts like Shift+F, R, and Backspace.
 *
 * @returns {JSX.Element}
 * - The rendered list component.
 */

const List = <T extends { id: number }, F extends keyof T>({
  header,
  checkboxDataCy,
  items,
  itemComposition,
  selectedItems,
  onClick,
  onDoubleClick,
  onEnterPressed,
  onSelectedItemsChanged,
  isLoading,
  forceLoading,
  skinSkeleton,
  emptyState,
  orderBy,
  onOrderByChanged,
  onNextPage,
  hasMoreItems,
  menu,
  displayMenuDiv,
  className,
  disableItemCompositionStyles,
  onMouseEnter,
  onMouseLeave,
  keyBoardShortcutActions,
  disableKeyboardShortcuts,
}: ListProps<T, F>): JSX.Element => {
  const containerRef = useRef<HTMLDivElement>(null);
  const isItemSelected = (item: T) => {
    return selectedItems.some((i) => item.id === i.id);
  };
  const container = document.getElementById('scrollableList');
  const isVerticalScrollbarVisible = container && container.scrollHeight > container.clientHeight;
  const isEmptyState = !hasMoreItems && items.length === 0 && !isLoading;
  const [idItemContextMenuOpen, setIdItemContextMenuOpen] = useState<number | null>(null);

  const skeletonData = new Array(25).fill(0).map(() => ({
    skeletonItem: skinSkeleton,
    columns: header.map((column) => column.width),
  }));

  const loader = <SkeletonLoader skeleton={skeletonData} />;

  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (containerRef.current && !containerRef.current.contains(e.target as Node)) {
        setIdItemContextMenuOpen(null);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    document.addEventListener('contextmenu', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.removeEventListener('contextmenu', handleClickOutside);
    };
  }, []);

  const handleNextPage = () => {
    onNextPage?.();
  };

  const unselectAllItems = () => {
    const changesToMake = selectedItems.map((item) => ({ props: item, value: false }));
    onSelectedItemsChanged(changesToMake);
  };

  const unselectAllItemsAndSelectOne = (props: T) => {
    const changesToMake = [...selectedItems.map((item) => ({ props: item, value: false })), { props, value: true }];
    onSelectedItemsChanged(changesToMake);
  };

  const executeClickOnSelectedItem = () => {
    const oneItemSelected = selectedItems.length === 1;
    if (oneItemSelected) {
      const selectedItem = selectedItems[0];
      onEnterPressed?.(selectedItem);
    }
  };

  const selectAllItems = () => {
    const notSelectedItems = _.difference(items, selectedItems);
    const changesToMake = notSelectedItems.map((item) => ({ props: item, value: true }));
    onSelectedItemsChanged(changesToMake);
  };

  const onTopSelectionCheckboxClick = () => {
    const areAllItemsSelected = selectedItems.length === items.length;

    if (areAllItemsSelected) {
      unselectAllItems();
    } else {
      selectAllItems();
    }
  };

  const onOrderableColumnClicked = (field: HeaderProps<T, F>) => {
    onCloseContextMenu();
    if (!field.orderable || !onOrderByChanged) return;

    const columnWasAlreadySelected = orderBy?.field === field.name;
    if (columnWasAlreadySelected) {
      onOrderByChanged({ field: field.name, direction: orderBy.direction === 'ASC' ? 'DESC' : 'ASC' });
    } else {
      onOrderByChanged({ field: field.name, direction: field.defaultDirection ?? 'ASC' });
    }
  };

  const handleKeyPress = (action: () => void) => {
    if (!disableKeyboardShortcuts) action();
  };

  const handleRKeyPressed = () => {
    keyBoardShortcutActions?.onRKeyPressed?.();
  };

  const handleBackspaceKeyPressed = () => {
    keyBoardShortcutActions?.onBackspaceKeyPressed?.();
  };

  useHotkeys(
    {
      'ctrl+a': () => handleKeyPress(selectAllItems),
      'meta+a': () => handleKeyPress(selectAllItems),
      esc: () => handleKeyPress(unselectAllItems),
      r: () => handleKeyPress(handleRKeyPressed),
      backspace: () => handleKeyPress(handleBackspaceKeyPressed),
      delete: () => handleKeyPress(handleBackspaceKeyPressed),
    },
    [items, selectedItems, disableKeyboardShortcuts],
  );

  const onItemClick = (itemClicked: T, e: React.MouseEvent<HTMLDivElement>) => {
    if (e.metaKey || e.ctrlKey) {
      onSelectedItemsChanged([{ props: itemClicked, value: !isItemSelected(itemClicked) }]);
    } else if (!isItemSelected(itemClicked)) {
      onClick?.(itemClicked);
    }
    setIdItemContextMenuOpen(null);
  };

  const onRightItemClick = (props: T, e: React.MouseEvent<HTMLDivElement>) => {
    e.preventDefault();
    if (!isItemSelected(props)) {
      unselectAllItemsAndSelectOne(props);
    }
    setIdItemContextMenuOpen(props.id);
  };

  const onCloseContextMenu = () => {
    setIdItemContextMenuOpen(null);
  };

  const handleThreeDotsButtonClick = (e: React.MouseEvent<HTMLButtonElement>, item: T) => {
    e.stopPropagation();

    setIdItemContextMenuOpen((prevId) => (prevId ? null : item.id));
    if (!isItemSelected(item)) {
      unselectAllItemsAndSelectOne(item);
    }
  };

  return (
    <div
      id="generic-list-component"
      className={`relative flex h-full flex-col overflow-x-hidden overflow-y-hidden ${className}`}
      ref={containerRef}
    >
      {!isEmptyState ? (
        <ListHeader
          selectedItems={selectedItems}
          onTopSelectionCheckboxClick={onTopSelectionCheckboxClick}
          items={items}
          header={header}
          orderBy={orderBy}
          onOrderableColumnClicked={onOrderableColumnClicked}
          menu={menu}
          displayMenuDiv={displayMenuDiv}
          isVerticalScrollbarVisible={isVerticalScrollbarVisible}
          checkboxDataCy={checkboxDataCy}
          onClose={onCloseContextMenu}
        />
      ) : null}

      {/* BODY */}
      <div id="scrollableList" className="flex h-full flex-col overflow-x-auto overflow-y-auto">
        {isEmptyState ? (
          emptyState
        ) : items.length > 0 && !forceLoading ? (
          <InfiniteScroll
            handleNextPage={handleNextPage}
            hasMoreItems={!!hasMoreItems}
            loader={loader}
            scrollableTarget="scrollableList"
          >
            {items.map((item, index) => (
              <ListItem<T>
                key={item.id}
                item={item}
                listIndex={index}
                isOpen={idItemContextMenuOpen === item.id}
                onClose={onCloseContextMenu}
                itemComposition={itemComposition}
                selected={isItemSelected(item)}
                onDoubleClick={onDoubleClick && (() => onDoubleClick(item))}
                onClick={(e) => onItemClick(item, e)}
                onClickContextMenu={(e) => onRightItemClick(item, e)}
                onThreeDotsButtonPressed={(e) => handleThreeDotsButtonClick(e, item)}
                columnsWidth={header.map((column) => column.width)}
                menu={menu}
                onSelectedChanged={(value) => onSelectedItemsChanged([{ props: item, value }])}
                disableItemCompositionStyles={disableItemCompositionStyles}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
                genericEnterKey={executeClickOnSelectedItem}
              />
            ))}
          </InfiniteScroll>
        ) : (
          <>{loader}</>
        )}

        {/* Click outside of the list to unselect all items */}
        {items.length > 0 && (
          <div
            data-testid="outside-click-element"
            className="h-full w-full py-6"
            onClick={unselectAllItems}
            onContextMenu={(e) => {
              e.preventDefault();
              unselectAllItems();
            }}
          />
        )}
      </div>
    </div>
  );
};

export default List;
