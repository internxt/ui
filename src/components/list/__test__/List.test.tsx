import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { afterEach, beforeAll, describe, expect, it, vi } from 'vitest';
import List from '../List';
import { HeaderProps } from '../ListHeader';

const mockOnClick = vi.fn();
const mockOnDoubleClick = vi.fn();
const mockOnSelectedItemsChanged = vi.fn();
const mockOnNextPage = vi.fn();
const mockOnOrderByChanged = vi.fn();

type Item = {
  id: string;
  name: string;
  age: number;
};

const header: HeaderProps<Item, keyof Item>[] = [
  { label: 'Name', width: '100px', name: 'name', orderable: true, defaultDirection: 'ASC' },
  { label: 'Age', width: '50px', name: 'age', orderable: true, defaultDirection: 'ASC' },
];

const items = [
  { id: 1, name: 'Item 1', age: 25 },
  { id: 2, name: 'Item 2', age: 30 },
];

const selectedItems = [{ id: 1, name: 'Item 1', age: 25 }];
const itemComposition = [(props: { name: string }) => <div>{props.name}</div>];

const emptyState = <div>No items to display</div>;

const menu = [
  { name: 'Option 1', action: () => {} },
  { name: 'Option 2', action: () => {} },
];

const defaultProps = {
  header,
  items,
  itemComposition,
  selectedItems,
  emptyState,
  menu,
  onSelectedItemsChanged: mockOnSelectedItemsChanged,
  onClick: mockOnClick,
  onDoubleClick: mockOnDoubleClick,
  onNextPage: mockOnNextPage,
  onOrderByChanged: mockOnOrderByChanged,
};

const selectedAllItemsMock = [
  {
    props: {
      age: 25,
      id: 1,
      name: 'Item 1',
    },
    value: true,
  },
  {
    props: {
      age: 30,
      id: 2,
      name: 'Item 2',
    },
    value: true,
  },
];

const unSelectedAllItemsMock = [
  {
    props: {
      age: 25,
      id: 1,
      name: 'Item 1',
    },
    value: false,
  },
  {
    props: {
      age: 30,
      id: 2,
      name: 'Item 2',
    },
    value: false,
  },
];

const renderList = (props = {}) => render(<List {...defaultProps} {...props} />);
beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (global as any).IntersectionObserver = class {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

describe('List component', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const list = renderList();
    expect(list).toMatchSnapshot();
  });

  it('renders the list header correctly', () => {
    const { getByText } = renderList();

    expect(getByText('Name')).toBeInTheDocument();
    expect(getByText('Age')).toBeInTheDocument();
  });

  it('renders items in the list', () => {
    const { getByText } = renderList();

    expect(getByText('Item 1')).toBeInTheDocument();
    expect(getByText('Item 2')).toBeInTheDocument();
  });

  it('calls onDoubleClick when an item is double clicked', () => {
    const { getByText } = renderList();

    const item = getByText('Item 1');
    fireEvent.doubleClick(item);

    expect(mockOnDoubleClick).toHaveBeenCalledWith(items[0]);
  });

  it('calls onSelectedItemsChanged when an item is selected', () => {
    const { getAllByRole } = renderList();

    const checkbox = getAllByRole('checkbox');
    fireEvent.click(checkbox[1]);

    expect(mockOnSelectedItemsChanged).toHaveBeenCalled();
  });

  it('calls onNextPage when intersection is detected and hasMoreItems is true', async () => {
    vi.stubGlobal(
      'IntersectionObserver',
      vi.fn((callback) => ({
        observe: () => {
          callback([{ isIntersecting: true }]);
        },
        unobserve: vi.fn(),
        disconnect: vi.fn(),
      })),
    );

    renderList({ hasMoreItems: true });

    await waitFor(() => {
      expect(mockOnNextPage).toHaveBeenCalled();
    });
  });

  it('not to be called onNextPage when intersection is detected and hasMoreItems is false', async () => {
    vi.stubGlobal(
      'IntersectionObserver',
      vi.fn((callback) => ({
        observe: () => {
          callback([{ isIntersecting: true }]);
        },
        unobserve: vi.fn(),
        disconnect: vi.fn(),
      })),
    );

    renderList({ hasMoreItems: false });

    await waitFor(() => {
      expect(mockOnNextPage).not.toHaveBeenCalled();
    });
  });

  it('renders empty state when there are no items and no more items to load', () => {
    const { getByText } = renderList({ items: [], hasMoreItems: false });

    const emptyState = getByText('No items to display');
    expect(emptyState).toBeInTheDocument();
  });

  it('calls onOrderByChanged when a sortable column header is clicked', () => {
    const { getByText } = renderList();

    const nameHeader = getByText('Name');
    fireEvent.click(nameHeader);

    expect(mockOnOrderByChanged).toHaveBeenCalled();
  });

  it('selects all items when select all checkbox is clicked', () => {
    const { getAllByRole } = renderList();

    const selectAllCheckbox = getAllByRole('checkbox');
    fireEvent.click(selectAllCheckbox[0]);

    expect(mockOnSelectedItemsChanged).toBeCalledWith(selectedAllItemsMock);
  });

  it('unselects all items when unselect all is clicked', () => {
    const { getAllByRole } = renderList({ selectedItems: items });

    const selectAllCheckbox = getAllByRole('checkbox');
    fireEvent.click(selectAllCheckbox[0]);

    expect(mockOnSelectedItemsChanged).toBeCalledWith(unSelectedAllItemsMock);
  });

  it('unselects all items when clicked outside the list', () => {
    const { container } = renderList({ selectedItems: items });

    const outsideListElement = container.getElementsByClassName('h-full w-full py-6')[0];
    expect(outsideListElement).toBeInTheDocument();

    fireEvent.click(outsideListElement);

    expect(mockOnSelectedItemsChanged).toBeCalledWith(unSelectedAllItemsMock);
  });

  it('handles keyboard shortcuts for select all (Ctrl+A)', () => {
    const { container } = renderList({ disableKeyboardShortcuts: false });

    fireEvent.keyDown(container, { key: 'a', ctrlKey: true });

    expect(mockOnSelectedItemsChanged).toBeCalledWith(selectedAllItemsMock);
  });

  it('does not trigger select all if keyboard shortcuts are disabled', () => {
    const { container } = renderList({ disableKeyboardShortcuts: true });

    fireEvent.keyDown(container, { key: 'a', ctrlKey: true });

    expect(mockOnSelectedItemsChanged).not.toHaveBeenCalled();
  });
});
