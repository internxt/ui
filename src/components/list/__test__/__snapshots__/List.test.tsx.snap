// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List component > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex h-full flex-col overflow-x-hidden overflow-y-hidden undefined"
        id="generic-list-component"
      >
        <div
          class="flex h-12 shrink-0 flex-row px-5"
        >
          <div
            class="flex h-full min-w-full flex-row items-center border-b border-gray-10"
          >
            <div
              class="flex h-full flex-row items-center justify-between pr-4"
            >
              <label
                class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
              >
                <div
                  class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-primary bg-primary"
                >
                  <svg
                    class="absolute -inset-px"
                    fill="none"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      fill="currentColor"
                      height="2"
                      rx="1"
                      width="12"
                      x="4"
                      y="9"
                    />
                  </svg>
                </div>
                <input
                  checked=""
                  class="base-checkbox h-0 w-0 appearance-none opacity-0"
                  readonly=""
                  type="checkbox"
                />
              </label>
            </div>
            <button
              class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  100px cursor-pointer hover:text-gray-80"
              data-cy="false"
            >
              <span
                data-cy="false"
              >
                Name
              </span>
            </button>
            <button
              class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  50px cursor-pointer hover:text-gray-80"
              data-cy="false"
            >
              <span
                data-cy="false"
              >
                Age
              </span>
            </button>
            <div
              class="flex h-full w-12 shrink-0"
            />
          </div>
        </div>
        <div
          class="flex h-full flex-col overflow-x-auto overflow-y-auto"
          id="scrollableList"
        >
          <div>
            <div
              class="group relative flex h-14 flex-row items-center pl-14 pr-5 bg-primary/10 text-gray-100 dark:bg-primary/20"
            >
              <div
                class="absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 opacity-100"
              >
                <label
                  class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
                >
                  <div
                    class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-primary bg-primary"
                    data-cy="driveListItemCheckbox0"
                  >
                    <svg
                      class="absolute -inset-px"
                      fill="none"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m8.85955 15c.38202 0 .67978-.1461.88202-.4551l5.23593-8.00557c.1461-.22472.2079-.43259.2079-.62922 0-.52809-.3933-.91011-.9326-.91011-.3708 0-.6011.13483-.8258.48876l-4.59554 7.24724-2.33146-2.8933c-.20787-.24719-.43259-.35955-.74719-.35955-.54495 0-.9382.38764-.9382.91575 0 .2359.07303.4438.27528.6741l2.89887 3.5113c.24158.2865.51124.4157.87079.4157z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                  <input
                    checked=""
                    class="base-checkbox h-0 w-0 appearance-none opacity-0"
                    readonly=""
                    type="checkbox"
                  />
                </label>
              </div>
              <div
                class="relative flex h-full shrink-0 flex-row items-center border-b border-primary/5 100px"
              >
                <div>
                  Item 1
                </div>
              </div>
              <div
                class="flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b border-primary/5"
              >
                <button
                  class="flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 text-gray-80 hover:bg-primary/10 focus-visible:bg-primary/10"
                >
                  <svg
                    fill="currentColor"
                    height="24"
                    viewBox="0 0 256 256"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M144,128a16,16,0,1,1-16-16A16,16,0,0,1,144,128ZM60,112a16,16,0,1,0,16,16A16,16,0,0,0,60,112Zm136,0a16,16,0,1,0,16,16A16,16,0,0,0,196,112Z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="group relative flex h-14 flex-row items-center pl-14 pr-5 focus-within:bg-gray-1 hover:bg-gray-1"
            >
              <div
                class="absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 false"
              >
                <label
                  class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
                >
                  <div
                    class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-gray-30 hover:border-gray-40"
                    data-cy="driveListItemCheckbox1"
                  />
                  <input
                    class="base-checkbox h-0 w-0 appearance-none opacity-0"
                    readonly=""
                    type="checkbox"
                  />
                </label>
              </div>
              <div
                class="relative flex h-full shrink-0 flex-row items-center border-b border-gray-5 100px"
              >
                <div>
                  Item 2
                </div>
              </div>
              <div
                class="flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b border-gray-5"
              >
                <button
                  class="flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 text-gray-60 hover:bg-gray-10 focus-visible:bg-gray-10"
                >
                  <svg
                    fill="currentColor"
                    height="24"
                    viewBox="0 0 256 256"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M144,128a16,16,0,1,1-16-16A16,16,0,0,1,144,128ZM60,112a16,16,0,1,0,16,16A16,16,0,0,0,60,112Zm136,0a16,16,0,1,0,16,16A16,16,0,0,0,196,112Z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
          <div
            class="h-full w-full py-6"
            data-testid="outside-click-element"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex h-full flex-col overflow-x-hidden overflow-y-hidden undefined"
      id="generic-list-component"
    >
      <div
        class="flex h-12 shrink-0 flex-row px-5"
      >
        <div
          class="flex h-full min-w-full flex-row items-center border-b border-gray-10"
        >
          <div
            class="flex h-full flex-row items-center justify-between pr-4"
          >
            <label
              class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
            >
              <div
                class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-primary bg-primary"
              >
                <svg
                  class="absolute -inset-px"
                  fill="none"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="currentColor"
                    height="2"
                    rx="1"
                    width="12"
                    x="4"
                    y="9"
                  />
                </svg>
              </div>
              <input
                checked=""
                class="base-checkbox h-0 w-0 appearance-none opacity-0"
                readonly=""
                type="checkbox"
              />
            </label>
          </div>
          <button
            class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  100px cursor-pointer hover:text-gray-80"
            data-cy="false"
          >
            <span
              data-cy="false"
            >
              Name
            </span>
          </button>
          <button
            class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  50px cursor-pointer hover:text-gray-80"
            data-cy="false"
          >
            <span
              data-cy="false"
            >
              Age
            </span>
          </button>
          <div
            class="flex h-full w-12 shrink-0"
          />
        </div>
      </div>
      <div
        class="flex h-full flex-col overflow-x-auto overflow-y-auto"
        id="scrollableList"
      >
        <div>
          <div
            class="group relative flex h-14 flex-row items-center pl-14 pr-5 bg-primary/10 text-gray-100 dark:bg-primary/20"
          >
            <div
              class="absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 opacity-100"
            >
              <label
                class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
              >
                <div
                  class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-primary bg-primary"
                  data-cy="driveListItemCheckbox0"
                >
                  <svg
                    class="absolute -inset-px"
                    fill="none"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m8.85955 15c.38202 0 .67978-.1461.88202-.4551l5.23593-8.00557c.1461-.22472.2079-.43259.2079-.62922 0-.52809-.3933-.91011-.9326-.91011-.3708 0-.6011.13483-.8258.48876l-4.59554 7.24724-2.33146-2.8933c-.20787-.24719-.43259-.35955-.74719-.35955-.54495 0-.9382.38764-.9382.91575 0 .2359.07303.4438.27528.6741l2.89887 3.5113c.24158.2865.51124.4157.87079.4157z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <input
                  checked=""
                  class="base-checkbox h-0 w-0 appearance-none opacity-0"
                  readonly=""
                  type="checkbox"
                />
              </label>
            </div>
            <div
              class="relative flex h-full shrink-0 flex-row items-center border-b border-primary/5 100px"
            >
              <div>
                Item 1
              </div>
            </div>
            <div
              class="flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b border-primary/5"
            >
              <button
                class="flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 text-gray-80 hover:bg-primary/10 focus-visible:bg-primary/10"
              >
                <svg
                  fill="currentColor"
                  height="24"
                  viewBox="0 0 256 256"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M144,128a16,16,0,1,1-16-16A16,16,0,0,1,144,128ZM60,112a16,16,0,1,0,16,16A16,16,0,0,0,60,112Zm136,0a16,16,0,1,0,16,16A16,16,0,0,0,196,112Z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="group relative flex h-14 flex-row items-center pl-14 pr-5 focus-within:bg-gray-1 hover:bg-gray-1"
          >
            <div
              class="absolute left-5 top-0 flex h-full w-0 flex-row items-center justify-start p-0 opacity-0 focus-within:opacity-100 group-hover:opacity-100 false"
            >
              <label
                class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
              >
                <div
                  class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-gray-30 hover:border-gray-40"
                  data-cy="driveListItemCheckbox1"
                />
                <input
                  class="base-checkbox h-0 w-0 appearance-none opacity-0"
                  readonly=""
                  type="checkbox"
                />
              </label>
            </div>
            <div
              class="relative flex h-full shrink-0 flex-row items-center border-b border-gray-5 100px"
            >
              <div>
                Item 2
              </div>
            </div>
            <div
              class="flex h-14 w-12 shrink-0 flex-col items-center justify-center border-b border-gray-5"
            >
              <button
                class="flex h-10 w-10 flex-col items-center justify-center rounded-md opacity-0 outline-none focus-visible:opacity-100 group-hover:opacity-100 text-gray-60 hover:bg-gray-10 focus-visible:bg-gray-10"
              >
                <svg
                  fill="currentColor"
                  height="24"
                  viewBox="0 0 256 256"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M144,128a16,16,0,1,1-16-16A16,16,0,0,1,144,128ZM60,112a16,16,0,1,0,16,16A16,16,0,0,0,60,112Zm136,0a16,16,0,1,0,16,16A16,16,0,0,0,196,112Z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="h-full w-full py-6"
          data-testid="outside-click-element"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
